### PR TITLE
wine64: improve Darwin support

### PR DESCRIPTION
**Pull request by reckenrode**
###### Motivation for this change

My ultimate goal is to get Wine, MoltenVK, and DXVK working on Darwin, so I can play _Final Fantasy XIV_ using a Nix-based Wine instead of CrossOver. MoltenVK and DXVK support will be coming in subsequent PRs once #158534 lands. I have a working derivation for DXVK, but I’m waiting on that before submitting.

This change itself does a few things to improve the build situation on Darwin:

1. Marks Linux-dependent settings in `packages.nix` as such;
2. Adds libusb, which is the basis of the new joystick stuff in Wine 7.0;
3. Adds libkrb5 and libunwind because those are packaged in Nix and seemed to have been missing;
4. Fixes `winemac.drv` to build with the 10.12 SDK;
5. Disables X11 on Darwin in favor of `winemac.drv`;
6. Adds some … lovely workarounds to allow MinGW to build PE DLLs on Darwin; and
7. Fixes the build of `wine64-preload` to not link against any libraries because that breaks it at least on aarch64-darwin under Rosetta 2.

Gstreamer support is currently dependent on #160252 to fix a build issue on Darwin.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by reckenrode** [comment]
I should also add that sane support and netapi probably should build on Darwin, but the packages required for those features don’t currently build. I don’t really need Samba support, but I do plan on fixing sane-backends, so I can use it with my ScanSnap scanner (it’s available in MacPorts, so I assume it can be made to build on Darwin but just hasn’t).

**Comment by reckenrode** [comment]
Result of `nixpkgs-review pr 160256` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>wine64 (wine64Packages.stableFull)</li>
    <li>wine64Packages.stagingFull</li>
    <li>wine64Packages.unstableFull</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>wine64Packages.base (wine64Packages.stable)</li>
    <li>wine64Packages.minimal</li>
    <li>wine64Packages.staging</li>
    <li>wine64Packages.unstable</li>
  </ul>
</details>

File: pkgs/misc/emulators/wine/base.nix [hunk 1]
```diff
@@ -15,9 +15,24 @@ let
   prevName = name;
   prevPlatforms = platforms;
   prevConfigFlags = configureFlags;
+  setupHook = makeSetupHook {
+    name = "darwin-mingw-hook";
+    substitutions = {
+      darwinSuffixSalt = stdenv.cc.suffixSalt;
+      mingwGccsSuffixSalts = map (gcc: gcc.suffixSalt) mingwGccs;
+    };
+  } ./setup-hook.sh;
```
Comment (bendlas) [inline pkgs/misc/emulators/wine/base.nix:24]: Maybe name this file `setup-hook-darwin.sh`, to make directory listing more useful.

**Review by bendlas — APPROVED** [review]
I can't comment on any darwin specifics, but overall, this seems like a reasonable step forward.

File: pkgs/misc/emulators/wine/base.nix (hunk 1 above)
Comment (7c6f434c) [reply pkgs/misc/emulators/wine/base.nix:24]: Arguably the variable name here could also include Darwin

**Comment by reckenrode** [comment]
> Result of `nixpkgs-review pr 160256` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
> 
> 3 packages failed to build:
> 4 packages built:

Pretty sure these failures are due to gst_all_1.gst-plugins-bad failing to build on Darwin currently. That’s fixed in #160252.

File: pkgs/misc/emulators/wine/base.nix (hunk 1 above)
Comment (reckenrode) [reply pkgs/misc/emulators/wine/base.nix]: I’ve updated the PR to do both. Thanks for the feedback!

**Comment by reckenrode** [comment]
I’ve rebased against master to fix the conflicts due to the refactor. Before I do that, I need to fix the linking problem with `wine64-preloader`. It’s linking against stuff, which it shouldn’t do, and I need to fix that to prevent everything from being broken on Darwin again.

**Comment by reckenrode** [comment]
I rebased and pushed updates to fix the build against master. Since #158534 landed, I also included Vulkan support for Darwin (no DXVK or anything like that yet).

Here’s a shot of the Windows build of vkQuake running under `wine64` on aarch64-darwin.
vkQuake Win

**Comment by reckenrode** [comment]
Last push was just a tweak to make MoltenVK more easily overridable by end users (in case they need patches or a different version for compatibility with their apps).

**Comment by 7c6f434c** [comment]
Looks unlikely to break Linux, and Darwin changes look reasonable apriori to me as macOS-avoider (and you tested them, after all)

**Comment by vcunat** [comment]
This never worked on Hydra: https://hydra.nixos.org/job/nixpkgs/trunk/wine64.x86_64-darwin/all

(I just noticed; I'm not interested in this combination.)

**Comment by reckenrode** [comment]
> This never worked on Hydra: https://hydra.nixos.org/job/nixpkgs/trunk/wine64.x86_64-darwin/all
> 
> (I just noticed; I'm not interested in this combination.)

MoltenVK requires Xcode to build on x86_64-darwin due needing a newer SDK. I can open a PR to disable building on Hydra, so that doesn’t happen. I can also disable Vulkan by default on Darwin for Wine, so users can get a cached build (but with no Vulkan support).

Once a newer SDK is available (at least 11.0), I can switch the MoltenVK derivation to use that. The derivation is written with using `xcbuild` in mind, but it only works on aarch64-darwin.

**Comment by reckenrode** [comment]
I opened #169782 to disable building MoltenVK on Hydra. I deferred on Wine64 for now because `wine64Packages.minimal` doesn’t pull in MoltenVK. This assumes Hydra won’t try to build `wine64Packages.base` or `wine64Packages.full` when one of the dependencies doesn’t build (I’m not super familiar with Hydra, so that may not be a good assumption).

**Comment by vcunat** [comment]
`meta.hydraPlaftorms` does not exclude transitively, contrary to `meta.platforms`.  It sounds counter-intuitively, but it's just implemented in a very different way.

**Comment by reckenrode** [comment]
> `meta.hydraPlaftorms` does not exclude transitively, contrary to `meta.platforms`. It sounds counter-intuitive, but it's just implemented in a very different way.

Ah, well then. I’ll push a change to Wine to disable Vulkan by default on Darwin to avoid the Hydra build failures.

**Pull request merged**